### ISSUE TITLE
Standard Event Tracking API

### ISSUE THREAD

**Issue by chpapa** [SkygearIO/features#33]
# Description

A standard event tracking API for simply store it in PostgreSQL, and redirect the events to 3rd party analytics tools.

# Design/Specification

Provide the following API for the module:
* `trackEvent`: `event`, `properties (JSON array)`, `revenue`, `currency`, `value`
* `trackPage`: `name`, `properties (JSON array)`, `path`, `referrer`, `query_str`, `title`, `url`
  * Can we automatically fill in path / referrer / query string / title and URL for JS SDK?
* `trackScreen`: `name`, `properties (JSON array)`

And should send the following information to others:
* Identify the current user
  * `userId`, `createdAt`, `email`, `phone`, `username`
  * profile, match with UserRecord if available: `address`, `age`, `avatar`, `birthday`, `desc`, `gender`, `firstname`, `lastname`, `fullname`, `title`, `url`
  * other custom fields, allow users to choose it or write cloud functions to customize it?
* Alias between anonymous and logged in users
* Any Tracked Events / Pages / Screens

On Portal, there should be a new session for users to:
* Enable / Disable storage of events to Database
* Enable / Config integrate with Google Analytics

# Related Issues

- Server Issues
- Client Issues
- Guides Issues

# Progress Tracker

- [ ] Specification Design Approval
- [ ] Write (code + tests + API docs) then get them merged. **Put All PR there**
  - [ ] Code
  - [ ] Minimal tests
  - [ ] Minimal API Docs
- [ ] Guides. **Put All PR there**
- [ ] Release
  - [ ] Update release notes
  - [ ] Release

# Advice

- Specification Design Approval
  - Once you get LGTM from another Skygear Core Team Member, you can check this checkbox. And apply the "workflow/design-complete" label.
- Coding
  - Use as many PRs as you need.  Write tests in the same or different PRs, as is convenient for you.
  - API doc should goes in the same PR with the code.
  - As each PR is merged, add a comment to this issue referencing the PRs.
  - When you are done with the code, apply the "workflow/code-complete" label.
- Guides
  - Write or modify guides and get them merged in https://github.com/skygeario/guides
  - When the PR of guides is merged, check this checkbox and apply the "workflow/guides-complete" label.

**Comment by chpapa** [comment]
@royuen I remember you have mention Oursky team have implemented similar plugins for numerous time. Can you take the lead to implement it when it's about the right time?

**Comment by chpapa** [comment]
@louischan-oursky FYI. The existing description and the notes on API design are just a brain dump of something I discussed with @royuen before. Feel free to overwrite it with what you would like to propose.

**Comment by chpapa** [comment]
@tensiuyan why you removed the WIP label? I think Louis is actually working on it...

**Comment by chpapa** [comment]
@rickmak @louischan-oursky so you two discussed and think the current spec is good? Shall we assume the specification are approved and it is design-complete already?

**Comment by rickmak** [comment]
No, none ping me for any discussion. I don't know who approval the design spec.

**Comment by louischan-oursky** [comment]
@rickmak the initial implementation of the plugin part is here ready for review. I am going to work on the three client (android, ios, javascript). Can you create three repositories for me?

**Comment by rickmak** [comment]
- https://github.com/SkygearIO/event_tracking-SDK-JS
- https://github.com/SkygearIO/event_tracking-SDK-iOS
- https://github.com/SkygearIO/event_tracking-SDK-Android

**Comment by chpapa** [comment]
Or use something like this? https://github.com/rudderlabs/rudder-server
